### DISCUSSION TITLE
How to use xct:Translate with {0} inside localized string?

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [xamarin/XamarinCommunityToolkit#743] [Q&A]
I have a label that displays app version like so "Version: 1.2". Previously I used this xmal to describe it:

```xaml
<Label Margin="0,12" 
                   FontSize="13" 
                   HorizontalOptions="Center" 
                   TextColor="Gray"
                   Text="{Binding ., Source={Static essentials:AppInfo.VersionString}, StringFormat={LN:Translate Version}}"/>
```
(`Version` string looks like this `Version: {0}`)

With `xct:Translate` I cannot do this since it returns `BindingBase` instead of `string`. 
What is the recommended way of doing this sort of thing using `xct:Translate`?

The only workaround I can see is to remove `{0}` from localized strings and split label into multiple spans, but that's not ideal since different localized values can have placeholder in different places (not in my case, but still).

**Comment by pictos** [comment]
@maxkoshevoi you can construct your string on C#. Is that possible in your scenario?

**Reply by maxkoshevoi** [reply]
Yes, I can create this string in view model. Hoped it could be done through xaml somehow

**Reply by pictos** [reply]
TBH not sure if you can do it directly from XCT, I'll try somethings next week and let you know. If I don't find a way I'll open an issue to implement something for it

**Reply by maxkoshevoi** [reply]
In the meanwhile created `LocalizedString` so that strings provided from ViewModel are automatically updated on language change: https://github.com/xamarin/XamarinCommunityToolkit/pull/750

**Reply by pictos** [reply]
@maxkoshevoi before implementing a new feature on enhancement is better to create an issue first, that way we can see and discuss the spec and proposal before doing a PR. That's mostly to avoid the contributors to work in something that could be refused.
Anyway, thank you for the help and I'll take a look at that PR ASAP.

**Reply by maxkoshevoi** [reply]
Thank you far taking a look at my PRs! Are discussions count as issues? Do I still need to create one #750?

**Comment by IeuanWalker** [comment]
@maxkoshevoi might not be ideal you could create a converter to handle it. The string in the resx is passed in as a command parameter

**Reply by maxkoshevoi** [reply]
I cannot wrap my head around how to make such convertor. Cause it need to return `Binding` that would react on `LocalizationResourceManager.Current.CurrectCulture` change, but contain only formatted string (instead of reference to `LocalizationResourceManager` like [current markup extension does](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/Extensions/TranslateExtension.shared.cs))

**Reply by IeuanWalker** [reply]
ahh good point.

I haven't used the language feature of this library yet but was thinking about updating our app to use it, as currently, we force an app restart to switch the language.

But this is how we use string format currently -
```xaml
Text="{Binding UserData.BillingPeriods.BillingPeriod[0].DebtBreakdown.AccountBalance, StringFormat={x:Static resx:AppResources.LblAccountInCredit}}"
````

Won't be able to update it to this library instead unless this is possible lol.

**Reply by IeuanWalker** [reply]
just for reference when I say force an app restart this is what I meant -
```c#
if (AppSettings.IsWelsh)
{
    CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("cy-GB");
    CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("cy-GB");
}
else
{
    CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("en-GB");
    CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("en-GB");
}

Application.Current.MainPage = new MainPage();
````

**Reply by maxkoshevoi** [reply]
If you move to it, here are couple of points:
1) Don't rely on `LocalizationResourceManager` yet. There is an issue that it doesn't persist culture to new pages #745. I created a fix for it (#749), but it's not reviewed or merged yet.
2) There are always some strings that are generated in the code and they need to be automatically updated on language change too. You can copy and use this [LocalizedString](https://github.com/xamarin/XamarinCommunityToolkit/pull/750) class in your project to achieve this until it is merged here. It will help to you with lack of xaml string formatting here. Your case will look something like this:
```cs
public LocalizedString Text { get; }
...
Text = new(() => string.Format(AppResources.LblAccountInCredit, UserData.BillingPeriods.BillingPeriod[0].DebtBreakdown.AccountBalance));
```

BTW interesting approach on reloading the application. I used Android's `activity.Recreate();` for that.

**Reply by IeuanWalker** [reply]
Ye been seeing your PR's coming through, nice work.
